### DISCUSSION TITLE
ERROR:  42501: permission denied for table members (RLS not working)

### DISCUSSION THREAD

**Discussion by mithunansivakumar** [supabase/supabase#28768] [Questions]
I'm implementing Row Level Security (RLS) on my Supabase database using the provided documentation. I have two tables: members and teams.

Here's my RLS policy:

```
create or replace function get_teams_for_user(user_id uuid)
returns setof bigint as $$
  select team_id from members where user_id = $1
$$ stable language sql security definer;

create policy "Team members can update team members if they belong to the team"
on members
for all using (
  team_id in (select get_teams_for_user(auth.uid()))
);
```
The goal is to restrict access to members records based on the user's team membership. However, when I execute SELECT * from members, I'm getting a denial even though the user is part of a team.

**Comment by GaryAustin1** [comment]
You mention a teams table but don't use it.  Should your function have the teams table in it?  

It might help to show your tables.

Also don't pass auth.uid(), instead just use it your function.  This will keep people from using the function to probe user ids.

**Comment by mithunansivakumar** [comment]
The tables:

![Screenshot 2024-08-20 at 18 06 51](https://github.com/user-attachments/assets/a29a7d85-5d4b-464a-aaaa-2359e19ac6fe)

![Screenshot 2024-08-20 at 18 07 02](https://github.com/user-attachments/assets/ec088c02-c16f-4762-b37a-de9c0da0041a)


And this is the example from supabase I am following to set the policy. 
![Screenshot 2024-08-20 at 18 07 24](https://github.com/user-attachments/assets/d32b22de-1364-42e4-a6cf-2672c4352cba)

**Reply by GaryAustin1** [reply]
That example does not seem right.    The function they show assumes your column name is user_id and it is the user's UUID.  You don't seem  to have that.  $1 means the parameter being passed.  

You don't label your tables but if the first one is members it does not have id as UUID and is not the user's UUID.  Auth.uid() will not work with any column you show.

If you are trying to allow the user to access the team table (is that the 2nd one?) then you would need to have your policy:

```sql
on teams
for all using (
  id in (select (get_teams_for_user(auth.uid()))
  )
  ```
  But the Supabase example is also bad.  It really should just be get_teams_for_user() and then you just use auth.uid() in the function itself, otherwise it is slower and insecure as the function can be used directly with anyone's user id.
  
  
  So you need to sort out where the auth user id will be stored in all of this as you don't seem to have it and you need to change the tables and names of columns to match your table names and columns.
